### PR TITLE
Add support for nested partition_key

### PR DESCRIPTION
**Pull request by zhonghui12**
*Issue #30*

*Description of changes:*
Add support for nested partition_key.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: kinesis/kinesis.go [hunk 1]
```diff
@@ -562,15 +562,32 @@ func (outputPlugin *OutputPlugin) randomString() string {
...
+	return nil
+}
+
 // getPartitionKey returns the value for a given valid key
 // if the given key is empty or invalid, it returns a random string
 func (outputPlugin *OutputPlugin) getPartitionKey(record map[interface{}]interface{}) string {
 	partitionKey := outputPlugin.partitionKey
+	partitionKeys := strings.Split(partitionKey, "->")
+	num := len(partitionKeys)
 	if partitionKey != "" {
```
Comment (PettitWesley) [inline kinesis/kinesis.go]: the check for `if partitionKey != "" {` should be first, right?

File: kinesis/kinesis.go [hunk 2]
```diff
@@ -562,15 +562,32 @@ func (outputPlugin *OutputPlugin) randomString() string {
...
+		for count, dataKey := range partitionKeys {
+			newRecord := getFromMap(dataKey, record)
+			_, ok := newRecord.(map[interface{}]interface{})
+			if ok {
+				record = newRecord.(map[interface{}]interface{})
+			} 
+			if count == num-1 {
+				value := stringOrByteArray(newRecord)
 				if value != "" {
 					if len(value) > partitionKeyMaxLength {
```
Comment (PettitWesley) [inline kinesis/kinesis.go]: I think the `if count == num - 1` should be at the top of the loop... that way you don't try to cast it to a map if its not needed

File: kinesis/kinesis.go [hunk 3]
```diff
@@ -562,15 +562,32 @@ func (outputPlugin *OutputPlugin) randomString() string {
...
-		for k, v := range record {
-			dataKey := stringOrByteArray(k)
-			if dataKey == partitionKey {
-				value := stringOrByteArray(v)
+		for count, dataKey := range partitionKeys {
+			newRecord := getFromMap(dataKey, record)
+			_, ok := newRecord.(map[interface{}]interface{})
+			if ok {
+				record = newRecord.(map[interface{}]interface{})
+			}
```
Comment (PettitWesley) [inline kinesis/kinesis.go]: I don't think this code properly accounts for error cases... please test those

What happens if the newRecord can not be cast to a map?

I think in that case you need break out of the loop and return the random partition key

File: kinesis/kinesis_test.go [hunk 4]
```diff
@@ -45,7 +45,7 @@ func newMockOutputPlugin(client *mock_kinesis.MockPutRecordsClient, isAggregate
 		stream:                "stream",
 		client:                client,
 		dataKeys:              "",
-		partitionKey:          "",
+		partitionKey:          "kubernetes->app",
```
Comment (PettitWesley) [inline kinesis/kinesis_test.go]: please add some more test cases:
- partition key with more than one "->"
- "normal" case where the partition key does not container "->"
- error case where the partition key is not found

File: kinesis/kinesis.go (hunk 3 above)
Comment (zhonghui12) [reply kinesis/kinesis.go]: yeah.. make sense. I lost this part after I updated the code. Will add it in next version

File: kinesis/kinesis.go (hunk 1 above)
Comment (zhonghui12) [reply kinesis/kinesis.go]: Make sense

File: kinesis/kinesis.go (hunk 2 above)
Comment (zhonghui12) [reply kinesis/kinesis.go]: yea. It is a good idea

File: kinesis/kinesis_test.go (hunk 4 above)
Comment (zhonghui12) [reply kinesis/kinesis_test.go]: Have already added more test cases for it

File: README.md [hunk 5]
```diff
@@ -11,7 +11,7 @@ If you think you’ve found a potential security issue, please do not post it in
...
+* `partition_key`: A partition key is used to group data by shard within a stream. A Kinesis Data Stream uses the partition key that is associated with each data record to determine which shard a given data record belongs to. For example, if your logs come from Docker containers, you can use container_id as the partition key, and the logs will be grouped and stored on different shards depending upon the id of the container they were generated from. As the data within a shard are coarsely ordered, you will get all your logs from one container in one shard roughly in order. Nested partition key is supported and you can use `->` to point to your target key which is nested under another key. If you don't set a partition key or put an invalid one, a random key will be generated, and the logs will be directed to random shards. If the partition key is invalid, the plugin will print an warning message.
```
Comment (PettitWesley) [inline README.md]: Add an extra sentence to show an example: "For example, your `partition_key` could be `"kubernetes->pod_name"`."

File: kinesis/kinesis.go [hunk 6]
```diff
@@ -562,22 +563,42 @@ func (outputPlugin *OutputPlugin) randomString() string {
...
 						value = value[0:partitionKeyMaxLength]
 					}
 					return value
 				}
 			}
+			_, ok := newRecord.(map[interface{}]interface{})
+			if ok {
+				record = newRecord.(map[interface{}]interface{})
+			} else {
+				logrus.Errorf("[kinesis %d] The partition key has a bad format; Will return a random string for it", outputPlugin.PluginID)
```
Comment (PettitWesley) [inline kinesis/kinesis.go]: Change to "The partition key could not be found in the record, using a random string instead"

Comment (zhonghui12) [reply kinesis/kinesis.go]: Updated the words

**Review by PettitWesley — APPROVED** [review]

**Pull request merged**